documentation/development-docs/patterns-for-standard-cmdlets.md

```diff
@@ -0,0 +1,996 @@
+                WriteObject(result);
+            }
+        }
+    }
+}
+```
+
+</p>
+</details>
+
+### `Remove-*` cmdlet
+
+All top-level resources should have a `Remove-*` cmdlet that allows users to delete a specific resource. The user can delete a resource by providing all identity properties, the resource id, or the object representation of the resource. This cmdlet should implement `SupportsShouldProcess` to allow users to provide the `-WhatIf` parameter and see what the result of executing the cmdlet is without the resource actually being deleted. This cmdlet should also implement the `-PassThru` parameter, which allows the user to receive output when no output would normally be provided.
+
+#### Parameter sets
+
+To enable the scenarios mentioned previously, the cmdlet will need three parameter sets:
+
+```
+Remove-AzureRmFoo -ResourceGroupName <String> -Name <String> [-PassThru] [-WhatIf] [-Confirm]
+
+Remove-AzureRmFoo -InputObject <PSFoo> [-PassThru] [-WhatIf] [-Confirm]
+
+Remove-AzureRmFoo -ResourceId <String> [-PassThru] [-WhatIf] [-Confirm]
+```
+
+The first parameter set has required `-ResourceGroupName` and `-Name` parameters, which allows the user to explicitly provide the identity properties of the resource that they want to delete. The second parameter has a required `-InputObject` parameter, which allows the user to pipe the result of the `Get-*` and `Set/Update-*` cmdlets to this cmdlet and delete the corresponding resource. The third parameter has a required `-ResourceId` parameter, which allows the user to delete a specific resource by resource id.
+
+#### C# example
+
+<details><summary>Click to expand example</summary>
+<p>
+
+```cs
+namespace Microsoft.Azure.Commands.Foo
+{
+    [Cmdlet(VerbsCommon.Remove, "AzureRmFoo", DefaultParameterSetName = "DeleteByNameParameterSet", SupportsShouldProcess = true), OutputType(typeof(bool))]
+    public class RemoveFooCommand : FooBaseCmdlet
+    {
+        [Parameter(Mandatory = true, ParameterSetName = "DeleteByNameParameterSet")]
+        [ResourceGroupCompleter]
+        [ValidateNotNullOrEmpty]
+        public string ResourceGroupName { get; set; }
+
+        [Parameter(Mandatory = true, ParameterSetName = "DeleteByNameParameterSet")]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = "DeleteByInputObjectParameterSet")]
+        [ValidateNotNull]
+        public PSFoo InputObject { get; set; }
+
+        [Parameter(Mandatory = true, ValueFromPipelineByPropertyName = true, ParameterSetName = "DeleteByResourceIdParameterSet")]
+        [ValidateNotNullOrEmpty]
+        public string ResourceId { get; set; }
+
+        [Parameter(Mandatory = false)]
+        public SwitchParameter PassThru { get; set; }
+
+        public override void ExecuteCmdlet()
+        {
+            if (this.IsParameterBound(c => c.InputObject))
```

Comment:
Why not use the param set name to figure out what's been passed in (like [here](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/Billing/Commands.Billing/Cmdlets/BillingPeriods/GetAzureRmBillingPeriod.cs#L39))?

Reply:
Same respond as above.